hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ProcessorTasklet.java

```diff
@@ -232,10 +237,17 @@ public void init() {
         } catch (Exception e) {
             throw sneakyThrow(e);
         }
+
+        if (!processor.isCooperative()) {
+            UserMetricsImp.set(userMetricsContext);
+        }
     }
 
     @Override @Nonnull
     public ProgressState call() {
+        if (processor.isCooperative()) {
+            UserMetricsImp.set(userMetricsContext);
```

Comment:
we should be able to avoid this by having a reference to the TL object, and updating a field on it.